Quarkus build with vertx-hazelcast fails after upgrade to 3.33.1

### Describe the bug

Hello,

after upgrading projects from Quarkus 3.27.3 to 3.33.1 I can't build the project. Maven package fails:

I use quarkus-vertx and [vertx-hazelcast](https://github.com/vert-x3/vertx-hazelcast/tree/4.5.25) as cluster manager. The later one provides [HazelcastClusterManager](https://github.com/vert-x3/vertx-hazelcast/blob/4.5.25/src/main/java/io/vertx/spi/cluster/hazelcast/HazelcastClusterManager.java) via SPI [VertxServiceProvider](https://github.com/vert-x3/vertx-hazelcast/blob/4.5.25/src/main/resources/META-INF/services/io.vertx.core.spi.VertxServiceProvider). quarkus-vertx uses this one in the io.quarkus.vertx.core.runtime.VertxCoreRecorder at build time pre processing.
There it is apptempted to serialize the HazelcastClusterManager. This seems to fail for the property `private volatile boolean active;`, wich is considered readonly. This is probably because no setter exists for it.

Question: Is this something that changed in the behaviour of the Quarkus build process? Is this a Quarkus bug? Or is this a vertx-hazelcast problem?



### Expected behavior

maven package succeeds.

### Actual behavior


Maven package leads to:
```
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.33.1:build (default) on project vertx-hazelcast-problem: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.deployment.steps.MainClassBuildStep#build threw an exception: java.lang.RuntimeException: Failed to record call to method public java.util.function.Supplier io.quarkus.vertx.core.runtime.VertxCoreRecorder.configureVertx(io.quarkus.runtime.LaunchMode,io.quarkus.runtime.ShutdownContext,java.util.List,java.util.List,java.util.List,java.util.concurrent.ExecutorService)
[ERROR] 	at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:493)
[ERROR] 	at io.quarkus.deployment.steps.MainClassBuildStep.writeRecordedBytecode(MainClassBuildStep.java:554)
[ERROR] 	at io.quarkus.deployment.steps.MainClassBuildStep.build(MainClassBuildStep.java:328)
[ERROR] 	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:735)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:898)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:242)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:1474)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: java.lang.RuntimeException: java.lang.RuntimeException: Cannot serialise field 'active' on object 'io.vertx.spi.cluster.hazelcast.HazelcastClusterManager@5f928fae' as the property is read only
[ERROR] 	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1370)
[ERROR] 	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstanceImpl(BytecodeRecorderImpl.java:985)
[ERROR] 	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstance(BytecodeRecorderImpl.java:599)
[ERROR] 	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1143)
[ERROR] 	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstanceImpl(BytecodeRecorderImpl.java:985)
[ERROR] 	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstance(BytecodeRecorderImpl.java:599)
[ERROR] 	at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:488)
```

### How to Reproduce?

Reproducer: https://github.com/hc42/quarkus-3.33-hazelcast-problem/tree/master
run maven pakage

### Output of `uname -a` or `ver`

Linux l-hhalfbrodt-moa 6.17.0-14-generic #14~24.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Jan 15 15:52:10 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "25.0.2" 2026-01-20 OpenJDK Runtime Environment (build 25.0.2+10-Ubuntu-124.04) OpenJDK 64-Bit Server VM (build 25.0.2+10-Ubuntu-124.04, mixed mode, sharing)

### Quarkus version or git rev

3.33.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.12

### Additional information

_No response_

--- a/extensions/vertx/deployment/src/main/java/io/quarkus/vertx/core/deployment/VertxCoreProcessor.java
+++ b/extensions/vertx/deployment/src/main/java/io/quarkus/vertx/core/deployment/VertxCoreProcessor.java
@@ -252,12 +252,12 @@ class VertxCoreProcessor {
                 .map(VertxOptionsConsumerBuildItem::getConsumer)
                 .toList();
 
-        // resolve the services at build time
-        List<VertxServiceProvider> vertxServiceProviders = loadServices(VertxServiceProvider.class);
-        List<VerticleFactory> verticleFactories = loadServices(VerticleFactory.class);
+        // resolve the service class names at build time, they will be instantiated at runtime
+        List<String> vertxServiceProviderClassNames = loadServiceClassNames(VertxServiceProvider.class);
+        List<String> verticleFactoryClassNames = loadServiceClassNames(VerticleFactory.class);
 
         Supplier<Vertx> vertx = recorder.configureVertx(launchMode.getLaunchMode(), shutdown, consumers,
-                vertxServiceProviders, verticleFactories, executorBuildItem.getExecutorProxy());
+                vertxServiceProviderClassNames, verticleFactoryClassNames, executorBuildItem.getExecutorProxy());
         syntheticBeans.produce(SyntheticBeanBuildItem.configure(Vertx.class)
                 .types(Vertx.class)
                 .scope(Singleton.class)
@@ -460,22 +460,15 @@ class VertxCoreProcessor {
         }
     }
 
-    private <T> List<T> loadServices(Class<T> serviceClass) throws IOException, ClassNotFoundException {
-        List<T> services = new ArrayList<>();
+    private List<String> loadServiceClassNames(Class<?> serviceClass) throws IOException, ClassNotFoundException {
+        List<String> serviceClassNames = new ArrayList<>();
         for (Class<?> serviceImplClass : ServiceUtil.classesNamedIn(Thread.currentThread().getContextClassLoader(),
                 "META-INF/services/" + serviceClass.getName())) {
             if (!QuarkusClassLoader.isClassPresentAtRuntime(serviceImplClass.getName())) {
                 continue;
             }
-            try {
-                services.add(serviceClass.cast(serviceImplClass.getDeclaredConstructor().newInstance()));
-            } catch (Exception e) {
-                throw new IllegalStateException(
-                        "Failed to instantiate declared " + serviceClass.getSimpleName() + " class: "
-                                + serviceImplClass.getName(),
-                        e);
-            }
+            serviceClassNames.add(serviceImplClass.getName());
         }
-        return services;
+        return serviceClassNames;
     }
 }

--- a/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java
+++ b/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java
@@ -114,13 +114,13 @@ public class VertxCoreRecorder {
     }
 
     public Supplier<Vertx> configureVertx(LaunchMode launchMode, ShutdownContext shutdown,
-            List<Consumer<VertxOptions>> customizers, List<VertxServiceProvider> vertxServiceProviders,
-            List<VerticleFactory> verticleFactories, ExecutorService executorProxy) {
+            List<Consumer<VertxOptions>> customizers, List<String> vertxServiceProviderClassNames,
+            List<String> verticleFactoryClassNames, ExecutorService executorProxy) {
         // The wrapper previously here to prevent the executor to be shutdown prematurely is moved to higher level to the io.quarkus.runtime.ExecutorRecorder
         QuarkusExecutorFactory.sharedExecutor = executorProxy;
         if (launchMode != LaunchMode.DEVELOPMENT) {
             vertx = new VertxSupplier(launchMode, vertxConfig.getValue(), customizers, threadPoolConfig.getValue(), shutdown,
-                    vertxServiceProviders, verticleFactories);
+                    vertxServiceProviderClassNames, verticleFactoryClassNames);
             // we need this to be part of the last shutdown tasks because closing it early (basically before Arc)
             // could cause problem to beans that rely on Vert.x and contain shutdown tasks
             shutdown.addLastShutdownTask(new Runnable() {
@@ -134,7 +134,7 @@ public class VertxCoreRecorder {
         } else {
             if (vertx == null) {
                 vertx = new VertxSupplier(launchMode, vertxConfig.getValue(), customizers, threadPoolConfig.getValue(),
-                        shutdown, vertxServiceProviders, verticleFactories);
+                        shutdown, vertxServiceProviderClassNames, verticleFactoryClassNames);
             } else if (vertx.v != null) {
                 tryCleanTccl();
             }
@@ -225,8 +225,8 @@ public class VertxCoreRecorder {
 
     public static Vertx initialize(VertxConfiguration conf, VertxOptionsCustomizer customizer,
             ThreadPoolConfig threadPoolConfig, ShutdownContext shutdown,
-            LaunchMode launchMode, List<VertxServiceProvider> vertxServiceProviders,
-            List<VerticleFactory> verticleFactories) {
+            LaunchMode launchMode, List<String> vertxServiceProviderClassNames,
+            List<String> verticleFactoryClassNames) {
 
         VertxOptions options = new VertxOptions();
 
@@ -249,6 +249,10 @@ public class VertxCoreRecorder {
                 return createVertxThread(target, name, worker, maxExecTime, maxExecTimeUnit, launchMode, nonDevModeTccl);
             }
         };
+        List<VertxServiceProvider> vertxServiceProviders = instantiateServices(vertxServiceProviderClassNames,
+                VertxServiceProvider.class);
+        List<VerticleFactory> verticleFactories = instantiateServices(verticleFactoryClassNames, VerticleFactory.class);
+
         if (conf != null && conf.cluster() != null && conf.cluster().clustered()) {
             CompletableFuture<Vertx> latch = new CompletableFuture<>();
             new VertxBuilder(options)
@@ -286,6 +290,20 @@ public class VertxCoreRecorder {
         LateBoundMDCProvider.setMDCProviderDelegate(VertxMDC.INSTANCE);
 
         return logVertxInitialization(vertx);
+    }
+
+    private static <T> List<T> instantiateServices(List<String> classNames, Class<T> serviceClass) {
+        List<T> services = new ArrayList<>();
+        for (String className : classNames) {
+            try {
+                Class<?> clazz = Thread.currentThread().getContextClassLoader().loadClass(className);
+                services.add(serviceClass.cast(clazz.getDeclaredConstructor().newInstance()));
+            } catch (Exception e) {
+                throw new IllegalStateException(
+                        "Failed to instantiate " + serviceClass.getSimpleName() + " class: " + className, e);
+            }
+        }
+        return services;
     }
 
     /**
@@ -721,28 +739,28 @@ public class VertxCoreRecorder {
         final VertxOptionsCustomizer customizer;
         final ThreadPoolConfig threadPoolConfig;
         final ShutdownContext shutdown;
-        final List<VertxServiceProvider> vertxServiceProviders;
-        final List<VerticleFactory> verticleFactories;
+        final List<String> vertxServiceProviderClassNames;
+        final List<String> verticleFactoryClassNames;
         Vertx v;
 
         VertxSupplier(LaunchMode launchMode, VertxConfiguration config, List<Consumer<VertxOptions>> customizers,
                 ThreadPoolConfig threadPoolConfig,
                 ShutdownContext shutdown,
-                List<VertxServiceProvider> vertxServiceProviders, List<VerticleFactory> verticleFactories) {
+                List<String> vertxServiceProviderClassNames, List<String> verticleFactoryClassNames) {
             this.launchMode = launchMode;
             this.config = config;
             this.customizer = new VertxOptionsCustomizer(customizers);
             this.threadPoolConfig = threadPoolConfig;
             this.shutdown = shutdown;
-            this.vertxServiceProviders = vertxServiceProviders;
-            this.verticleFactories = verticleFactories;
+            this.vertxServiceProviderClassNames = vertxServiceProviderClassNames;
+            this.verticleFactoryClassNames = verticleFactoryClassNames;
         }
 
         @Override
         public synchronized Vertx get() {
             if (v == null) {
-                v = initialize(config, customizer, threadPoolConfig, shutdown, launchMode, vertxServiceProviders,
-                        verticleFactories);
+                v = initialize(config, customizer, threadPoolConfig, shutdown, launchMode, vertxServiceProviderClassNames,
+                        verticleFactoryClassNames);
             }
             return v;
         }

